clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java

```diff
@@ -2416,10 +2424,22 @@ public void testIncrementalAlterConfigs()  throws Exception {
                     .setErrorCode(Errors.INVALID_REQUEST.code())
                     .setErrorMessage("Config value append is not allowed for config"));
 
+            responseData.responses().add(new AlterConfigsResourceResponse()
+                    .setResourceName("topic2")
+                    .setResourceType(ConfigResource.Type.TOPIC.id())
+                    .setErrorCode(Errors.NOT_CONTROLLER.code())
+                    .setErrorMessage("Topic resource change must be sent to the controller"));
+
             env.kafkaClient().prepareResponse(new IncrementalAlterConfigsResponse(responseData));
 
+            // Upon handling NOT_CONTROLLER exception, admin client needs to refresh the metadata.
+            MetadataResponse controllerNodeResponse = MetadataResponse.prepareResponse(env.cluster().nodes(),
+                env.cluster().clusterResource().clusterId(), 1, Collections.emptyList());
+            env.kafkaClient().prepareResponse(controllerNodeResponse);
```

Comment:
It may be worth using `prepareResponseFrom` in the test cases to verify that requests go to the expected brokers. I would also extend the test cases to cover the logic which create one or multiple requests based on wether `shouldValidateOnly` is set.